File: rust/src/leetcode/1.two-sum.rs
/*
 * @lc app=leetcode id=1 lang=rust
 *
 * [1] Two Sum
 *
 * https://leetcode.com/problems/two-sum/description/
 *
 * algorithms
 * Easy (45.90%)
 * Likes:    18231
 * Dislikes: 650
 * Total Accepted:    3.6M
 * Total Submissions: 7.9M
 * Testcase Example:  '[2,7,11,15]\n9'
 *
 * Given an array of integers nums and an integer target, return indices of the
 * two numbers such that they add up to target.
 *
 * You may assume that each input would have exactly one solution, and you may
 * not use the same element twice.
 *
 * You can return the answer in any order.
 *
 *
 * Example 1:
 *
 *
 * Input: nums = [2,7,11,15], target = 9
 * Output: [0,1]
 * Output: Because nums[0] + nums[1] == 9, we return [0, 1].
 *
 *
 * Example 2:
 *
 *
 * Input: nums = [3,2,4], target = 6
 * Output: [1,2]
 *
 *
 * Example 3:
 *
 *
 * Input: nums = [3,3], target = 6
 * Output: [0,1]
 *
 *
 *
 * Constraints:
 *
 *
 * 2 <= nums.length <= 10^3
 * -10^9 <= nums[i] <= 10^9
 * -10^9 <= target <= 10^9
 * Only one valid answer exists.
 *
 *
 */

struct Solution;
// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut h: HashMap<i32, i32> = HashMap::new();
        for (idx, i) in nums.iter().enumerate() {
            let left = target - i;
            match h.get(&left) {
                Some(v) => return vec![*v, idx as i32],
                None => h.insert(*i, idx as i32),
            };
        }
        return vec![];
    }
}
// @lc code=end

File: cpp/1.two-sum.cpp
/**
 * @lc app=leetcode id=1 lang=cpp
 *
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
   public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        unordered_map<int, int> mapping;

        for (int i = 0; i < nums.size(); i++) {
            mapping[nums[i]] = i;
        }
        for (int i = 0; i < nums.size(); i++) {
            int searched = target - nums[i];
            if (mapping.find(searched) != mapping.end() &&
                i != mapping[searched]) {
                result.push_back(i);
                result.push_back(mapping[searched]);
                break;
            }
        }
        return result;
    }
};
// @lc code=end
